File: android/src/main/java/com/lebrislo/bluetooth/mesh/plugin/VendorPluginCall.kt
package com.lebrislo.bluetooth.mesh.plugin

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import no.nordicsemi.android.mesh.transport.MeshMessage
import no.nordicsemi.android.mesh.transport.VendorModelMessageStatus

/**
 * This class is used to generate a response for a Vendor plugin call.
 */
class VendorPluginCall(
    val modelId: Int,
    val meshOperation: Int,
    val meshOperationCallback: Int,
    val meshAddress: Int,
    call: PluginCall
) :
    BasePluginCall(call) {
    companion object {
        /**
         * Generates a response for a Vendor plugin call.
         *
         * @param meshMessage Mesh message.
         */
        @JvmStatic
        fun generateVendorPluginCallResponse(meshMessage: MeshMessage): JSObject {
            meshMessage as VendorModelMessageStatus
            val result = JSObject()
            result.put("src", meshMessage.src)
            result.put("dst", meshMessage.dst)
            result.put("opcode", meshMessage.opCode)
            result.put("vendorModelId", meshMessage.modelIdentifier)
            val data = JSArray()
            meshMessage.parameters.forEach {
                data.put(it.toInt())
            }
            result.put("data", data)
            return result
        }
    }
}

File: ios/Sources/BluetoothMeshPlugin/plugin/VendorPluginCall.swift
//
//  VendorPluginCall.swift
//  BluetoothMeshPlugin
//
//  Created by LE BRIS Loris on 14/11/2025.
//

import Capacitor
import Foundation
import NordicMesh

public class VendorPluginCall: BasePluginCall {
    let modelId: UInt32
    let meshOperation: UInt32
    let meshOperationCallback: UInt32
    let meshAddress: UInt16
    let pluginCall: CAPPluginCall

    init(
        _ modelId: UInt32,
        _ meshOperation: UInt32,
        _ meshOperationCallback: UInt32,
        _ meshAddress: UInt16,
        _ call: CAPPluginCall
    ) {
        self.modelId = modelId
        self.meshOperation = meshOperation
        self.meshOperationCallback = meshOperationCallback
        self.meshAddress = meshAddress
        self.pluginCall = call
        super.init(call: call)
    }

    static func generateVendorPluginCallResponse(_ response: RoutedMeshMessage) -> PluginCallResultData {
        var result = PluginCallResultData()
        result["src"] = response.src
        result["dst"] = response.dst
        result["opcode"] = response.message.opCode
        result["vendorModelId"] = (response.message as? VendorResponse)?.companyIdentifier
        result["data"] = response.message.parameters
        return result
    }
}
